NDTensors/src/lib/SymmetrySectors/src/sector_product.jl

```diff
@@ -0,0 +1,240 @@
+# This files defines a structure for Cartesian product of 2 or more fusion sectors
+# e.g. U(1)×U(1), U(1)×SU2(2)×SU(3)
+
+using BlockArrays: blocklengths
+using ..LabelledNumbers: LabelledInteger, label, labelled, unlabel
+using ..GradedAxes: AbstractGradedUnitRange, GradedAxes, dual
+
+# =====================================  Definition  =======================================
+struct SectorProduct{Sectors} <: AbstractSector
+  arguments::Sectors
+  global _SectorProduct(l) = new{typeof(l)}(l)
+end
+
+SectorProduct(c::SectorProduct) = _SectorProduct(arguments(c))
+
+arguments(s::SectorProduct) = s.arguments
+
+# =================================  Sectors interface  ====================================
+function SymmetryStyle(T::Type{<:SectorProduct})
+  return arguments_symmetrystyle(arguments_type(T))
+end
+
+function quantum_dimension(::NotAbelianStyle, s::SectorProduct)
+  return mapreduce(quantum_dimension, *, arguments(s))
+end
+
+# use map instead of broadcast to support both Tuple and NamedTuple
+GradedAxes.dual(s::SectorProduct) = SectorProduct(map(dual, arguments(s)))
+
+function trivial(type::Type{<:SectorProduct})
+  return SectorProduct(arguments_trivial(arguments_type(type)))
+end
+
+# ===================================  Base interface  =====================================
+function Base.:(==)(A::SectorProduct, B::SectorProduct)
+  return arguments_isequal(arguments(A), arguments(B))
+end
+
+function Base.show(io::IO, s::SectorProduct)
+  (length(arguments(s)) < 2) && print(io, "sector")
+  print(io, "(")
+  symbol = ""
+  for p in pairs(arguments(s))
+    print(io, symbol)
+    sector_show(io, p[1], p[2])
+    symbol = " × "
+  end
+  return print(io, ")")
+end
+
+sector_show(io::IO, ::Int, v) = print(io, v)
+sector_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
+
+function Base.isless(s1::SectorProduct, s2::SectorProduct)
+  return arguments_isless(arguments(s1), arguments(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for SectorProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+arguments_type(::Type{<:SectorProduct{T}}) where {T} = T
+
+arguments_maybe_insert_unspecified(s1, ::Any) = s1
+function sym_arguments_maybe_insert_unspecified(s1, s2)
+  return arguments_maybe_insert_unspecified(s1, s2),
+  arguments_maybe_insert_unspecified(s2, s1)
+end
+
+function arguments_isequal(s1, s2)
+  isempty(s1) && return _arguments_isequal(empty(s2), s2)
+  isempty(s2) && return _arguments_isequal(s1, empty(s1))
+  return _arguments_isequal(s1, s2)
+end
+
+function _arguments_isequal(s1, s2)
+  return ==(sym_arguments_maybe_insert_unspecified(s1, s2)...)
+end
```

Comment:
Can you get rid of `_arguments_isequal` using a similar code pattern to the one used in `arguments_isless`?